pyomo/repn/plugins/nl_writer.py

```diff
@@ -0,0 +1,2128 @@
+
+@WriterFactory.register(
+    'nl_v2', 'Generate the corresponding AMPL NL file (version 2).')
+class NLWriter(object):
+    CONFIG = ConfigBlock('nlwriter')
+    CONFIG.declare('show_section_timing', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Print timing after writing each section of the NL file',
+    ))
+    CONFIG.declare('skip_trivial_constraints', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Skip writing constraints whose body is constant'
+    ))
+    CONFIG.declare('file_determinism', ConfigValue(
+        default=FileDeterminism.ORDERED,
+        domain=InEnum(FileDeterminism),
+        description='How much effort to ensure file is deterministic',
+        doc="""
+        How much effort do we want to put into ensuring the
+        NL file is written deterministically for a Pyomo model:
+            NONE (0) : None
+            ORDERED (1): rely on underlying component ordering (default)
+            SORT_INDICES (2) : sort keys of indexed components
+            SORT_SYMBOLS (3) : sort keys AND sort names (over declaration order)
+        """
+    ))
+    CONFIG.declare('symbolic_solver_labels', ConfigValue(
+        default=False,
+        domain=bool,
+        description='Write the corresponding .row and .col files',
+    ))
+    CONFIG.declare('export_nonlinear_variables', ConfigValue(
+        default=None,
+        domain=list,
+        description='Extra variables to include in NL file',
+        doc="""
+        List of variables to ensure are in the NL file (even if they
+        don't appear in any constraints)."""
+    ))
+    CONFIG.declare('row_order', ConfigValue(
+        default=None,
+        description='Preferred constraint ordering',
+        doc="""
+        List of constraints in the order that they should appear in the
+        NL file.  Note that this is only a suggestion, as the NL writer
+        will move all nonlinear constraints before linear ones
+        (preserving row_order within each group)."""
+    ))
+    CONFIG.declare('column_order', ConfigValue(
+        default=None,
+        description='Preferred variable ordering',
+        doc="""
+        List of variables in the order that they should appear in the NL
+        file.  Note that this is only a suggestion, as the NL writer
+        will move all nonlinear variables before linear ones, and within
+        nonlinear variables, variables appearing in both objectives and
+        constraints before variables appearing only in constraints,
+        which appear before variables appearing only in objectives.
+        Within each group, continuous variables appear before discrete
+        variables.  In all cases, column_order is preserved within each
+        group."""
+    ))
+
+    def __init__(self):
+        self.config = self.CONFIG()
+
+    def __call__(self, model, filename, solver_capability, io_options):
```

Comment:
Please document the arguments and return value for this method.

Reply:
Can this method please also return the ordering of the variables and the constraints (or instructions on how to access this.

Reply:
I will add some documentation as soon as I can figure out what the arguments are.  ;)  This API can't really be changed: it is here for backwards compatibility with the old writers.  The more I work with core, the more I would like to deprecate most uses of `__call__()`.

Reply:
Haha!

Can we create a public API and then wrap it for use with the `WriterFactory` and backwards compatibility (similar to what we are doing for Appsi)?

Reply:
Oh! My bad. I see. I just want the `write` method, and the variable constraint ordering is in `NLWriterInfo`? Maybe I really just want to you document the `write` method and `NLWriterInfo`. I'm not sure I care about documenting `__call__`.